### ISSUE TITLE
Allow CA1859 to also be reported for internal, protected or public members

### ISSUE THREAD

**Issue by drieseng** [dotnet/roslyn-analyzers#6564]
### Analyzer

**Diagnostic ID**: [CA1859]

### Describe the improvement

Allow the **CA1859** diagnostic to also be reported for internal, protected or public members.
Changing the return type of members with that visibility could/would of course by considered a breaking change, which is why this should be opt-in.

Right now, **CA1859** is only reported for private members.

**Example:**
Given the following code fragment:
```c#
using System.Collections.Generic;

public class X
{
    private static IList<string> GetListPrivate()
    {
        return new List<string>();
    }

    internal static IList<string> GetListInternal()
    {
        return new List<string>();
    }

    protected static IList<string> GetListProtected()
    {
        return new List<string>();
    }

    public static IList<string> GetListPublic()
    {
        return new List<string>();
    }
}
```

and the following .editorconfig:

```
[*.cs]
dotnet_diagnostic.CA1859.severity = error
dotnet_code_quality.CA1859.api_surface = private, internal, public
```

**CA1859** is only reported for the `GetListPrivate()` method.

### Describe suggestions on how to achieve the rule

I propose to use the **dotnet_code_quality.CA1859.api_surface** option to select the API surface to run against.
By default, **CA1859** should only be reported for private (or private and internal) methods.

**Comment by drieseng** [comment]
Let me know if you want me to add this as a comment to #6403.

**Comment by drieseng** [comment]
**Update:**
This appears to be an IDE issue. When I build the project, **CA1859** is effectively reported for all methods, but the code editor in Visual Studio 2022 (version 17.5.3) only reports **CA1859** for the private method.

**Comment by mavasani** [comment]
> **Update:** This appears to be an IDE issue. When I build the project, **CA1859** is effectively reported for all methods, but the code editor in Visual Studio 2022 (version 17.5.3) only reports **CA1859** for the private method.

This is by design. **CA1859** for non-private symbols is reported via CompilationEnd action, which needs whole project analysis and is expensive to execute during live analysis. You can enable Entire solution background analysis to enable this by default for live analysis: https://learn.microsoft.com/en-us/visualstudio/code-quality/configure-live-code-analysis-scope-managed-code?view=vs-2022#custom-analysis-scope
